cdc/redo/writer/writer.go

```diff
@@ -99,72 +105,79 @@ type LogWriter struct {
-			Dir:               cfg.Dir,
-			ChangeFeedID:      cfg.ChangeFeedID,
-			CaptureID:         cfg.CaptureID,
-			FileType:          common.DefaultDDLLogFileType,
-			CreateTime:        cfg.CreateTime,
-			MaxLogSize:        cfg.MaxLogSize,
-			FlushIntervalInMs: cfg.FlushIntervalInMs,
-			S3Storage:         cfg.S3Storage,
-			S3URI:             cfg.S3URI,
-		}
-		logWriter = &LogWriter{
-			cfg: cfg,
-		}
-		logWriter.rowWriter, errInit = NewWriter(ctx, rowCfg)
-		if errInit != nil {
-			return
-		}
-		logWriter.ddlWriter, errInit = NewWriter(ctx, ddlCfg)
-		if errInit != nil {
-			return
+	initLock.Lock()
+	defer initLock.Unlock()
+
+	if v, ok := logWriters[cfg.ChangeFeedID]; ok {
+		// if cfg changed or already closed need create a new LogWriter
+		if cfg.String() == v.cfg.String() && !v.isStopped() {
+			return v, nil
 		}
+	}
 
-		// since the error will not block write log, so keep go to the next init process
-		err := logWriter.initMeta(ctx)
+	var err error
+	var logWriter *LogWriter
+	rowCfg := &FileWriterConfig{
+		Dir:               cfg.Dir,
+		ChangeFeedID:      cfg.ChangeFeedID,
+		CaptureID:         cfg.CaptureID,
+		FileType:          common.DefaultRowLogFileType,
+		CreateTime:        cfg.CreateTime,
+		MaxLogSize:        cfg.MaxLogSize,
+		FlushIntervalInMs: cfg.FlushIntervalInMs,
+		S3Storage:         cfg.S3Storage,
+		S3URI:             cfg.S3URI,
+	}
+	ddlCfg := &FileWriterConfig{
+		Dir:               cfg.Dir,
+		ChangeFeedID:      cfg.ChangeFeedID,
+		CaptureID:         cfg.CaptureID,
+		FileType:          common.DefaultDDLLogFileType,
+		CreateTime:        cfg.CreateTime,
+		MaxLogSize:        cfg.MaxLogSize,
+		FlushIntervalInMs: cfg.FlushIntervalInMs,
+		S3Storage:         cfg.S3Storage,
+		S3URI:             cfg.S3URI,
+	}
+	logWriter = &LogWriter{
+		cfg: cfg,
+	}
+	logWriter.rowWriter, err = NewWriter(ctx, rowCfg)
+	if err != nil {
+		return nil, err
+	}
+	logWriter.ddlWriter, err = NewWriter(ctx, ddlCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// since the error will not block write log, so keep go to the next init process
+	err = logWriter.initMeta(ctx)
+	if err != nil {
+		log.Warn("init redo meta fail",
+			zap.String("change feed", cfg.ChangeFeedID),
+			zap.Error(err))
+	}
+	if cfg.S3Storage {
+		logWriter.storage, err = common.InitS3storage(ctx, cfg.S3URI)
 		if err != nil {
-			log.Warn("init redo meta fail",
-				zap.String("change feed", cfg.ChangeFeedID),
-				zap.Error(err))
+			return nil, err
 		}
-		if cfg.S3Storage {
-			var s3storage storage.ExternalStorage
-			s3storage, errInit = common.InitS3storage(ctx, cfg.S3URI)
-			if errInit != nil {
-				return
-			}
-			logWriter.storage = s3storage
+	}
+	// close previous writer
+	if v, ok := logWriters[cfg.ChangeFeedID]; ok {
+		err = v.Close()
+		if err != nil {
+			return nil, err
 		}
-
-		go logWriter.runGC(ctx)
-	})
-	if errInit != nil {
-		return nil, errInit
 	}
+	logWriters[cfg.ChangeFeedID] = logWriter
+	go logWriter.runGC(ctx)
```

Comment:
If a changefeed is paused, the `ctx` will be cancelled. After the changefeed is resumed, will the `runGC` goroutine run again?